File: leetcode/DynamicProgram/leetcode-0123/123.go
package dynamicprogram

func maxProfit(prices []int) int {
	if prices == nil || len(prices) == 0 {
		return 0
	}

	global := make([]int, 3)
	local := make([]int, 3)

	for i := 0; i < len(prices)-1; i++ {
		diff := prices[i+1] - prices[i]
		for j := 2; j >= 1; j-- {
			local[j] = max(global[j-1]+max(0, diff), local[j]+diff)
			global[j] = max(global[j], local[j])
		}
	}
	return global[2]
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: leetcode/DynamicProgram/leetcode-0123/123.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) {
            return 0;
        }
        int n = prices.length;
        int[][] global = new int[n][3];
        int[][] local = new int[n][3];
        
        for (int i = 1; i < n; i++) {
            int diff = prices[i] - prices[i - 1];
            for (int j = 1; j <= 2; j++) {
                local[i][j] = Math.max(global[i - 1][j - 1] + Math.max(diff, 0), local[i - 1][j] + diff);
                global[i][j] = Math.max(global[i - 1][j], local[i][j]);
            }
        }
        return global[n - 1][2];
    }
}